internal/ledger/compare.go

```diff
@@ -0,0 +1,413 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ledger
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric/common/ledger/util"
+	"github.com/hyperledger/fabric/core/ledger/kvledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/privacyenabledstate"
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/pkg/errors"
+)
+
+// Compare - Compares two ledger snapshots and outputs the differences in snapshot records
+// This function will overwrite the file at outputPath if it already exists
+func Compare(snapshotDir1 string, snapshotDir2 string, outputFile string) (err error) {
+	// Check the hashes between two files
+	hashPath1 := filepath.Join(snapshotDir1, kvledger.SnapshotSignableMetadataFileName)
+	hashPath2 := filepath.Join(snapshotDir2, kvledger.SnapshotSignableMetadataFileName)
+
+	equal, err := snapshotsComparable(hashPath1, hashPath2)
+	if err != nil {
+		return err
+	}
+
+	if equal {
+		return errors.New("both snapshots public state hashes are same. Aborting compare")
+	}
+
```

Comment:
It seems it will always stop here and not proceed down to the actual comparisons below.

If _snapshot_signable_metadata.json don't match, it will have returned an error and exit (it shouldn't return...it should proceed to the comparisons).

If _snapshot_signable_metadata.json are equal, it always returns an error and exits (I think it is correct to exit in this case, but I wouldn't call it an error, you could just return nil for error when they match).

Reply:
Regardless of whether hashes are same or different, we should print the result at this point, and mention that all records will now be compared and it may take some time if there are many records to compare.

Reply:
This comparison uses the snapshot hash which is generated when the snapshot is created from the ledger, not the block hash. When the snapshot hash is equal, we know there are no differences between the snapshots.